### ISSUE TITLE
Add a portal for device access

### ISSUE THREAD

**Issue by matthiasclasen** [flatpak/xdg-desktop-portal#13]
cdrom, dvd, joystick, etc. Need to treat them case-by-case.

Totem case: need full dvd access. Problem is that we can only set up device access at startup time, since we want to drop privileges. VLC also wants full optical device access, which is considered to be a showstopper for publishing VLC as a Flatpak.

joystick: one idea would be to have the compositor forward events in a somewhat transparent way. This has the advantage that we can subject these events to focus. But there's an open question if we need any custom ioctls. Need to talk to domain experts: sdl and libinput maintainers.

Richard wants to have usb access to update firmware. We tentatively agreed to just have a way to bind mount /dev in the metadata, for cases like this. Full device access is not quite as scary as it sounds since permissions still limit what is accessible. 

Unfortunately, udev is not stable, so we can't just use libudev in the runtime. There was some idea  that we might have a mini-udev wrapper for the host udev.

**Comment by nedrichards** [comment]
Another case here would be scanner access, we're seeing people having to [resort to horrid hacks](https://github.com/flathub/flathub/pull/157) over on flathub. Although perhaps this should be a case of its own, like Printers.

**Comment by hadess** [comment]
Scanners should go through saned. If saned isn't good enough, then it can be enhanced. Being able to have 2 apps that talk to scanners opened at the same time seems like a low bar. I encountered the problem on my first day having my own scanner.

**Comment by jflesch** [comment]
The tricky thing with scanners are proprietary drivers. If an user has one installed on the host system, the application must be able to use it. The only solution that I can think of is starting Saned on the host system and make it listen on the loopback (port tcp/6566).

**Comment by dsboger** [comment]
For scanners, would it be a possible/desirable solution to provide a generic GTK+ widget/dialog, maybe in xdg-desktop-portal-gtk, maybe also to GTK+ itself, similar to file chooser and printing?

**Comment by jflesch** [comment]
[In my case](https://openpaper.work), not desirable at all. Scanning is fully integrated the UI, and I doubt a Gtk widget would cut it. Same goes for simple-scan and probably others :/

**Comment by jflesch** [comment]
Also, it would assume using GTK (or even a GUI). But scanning is also possible from the shell for scripting purposes.

**Comment by hadess** [comment]
> For scanners, would it be a possible/desirable solution to provide a generic GTK+ widget/dialog, maybe in xdg-desktop-portal-gtk, maybe also to GTK+ itself, similar to file chooser and printing?

It's not within scope.

**Comment by dsboger** [comment]
> It's not within scope.

@hadess could you elaborate a little? Scanner access was mentioned in this issue, but apps don't actually access the device, IIUC. Sane provides an abstraction service. Could/should that be embedded/converted into a portal?

**Comment by jflesch** [comment]
> From what I could see, both Paperwork and Simple Scan do exactly the same thing in the exact same way: there is a button to trigger a scan and a preview widget to show the ongoing capture.

Hm, yes and no. Paperwork always scan the maximum scan area possible, and while the scan is going, it highlight the part of the image it will crop out. I need to check but I think Simple-scan doesn't do that.

Other than that, I guess you're right. Both are quite similar visually.

However that it is not always the most common way to scan and display scan results. For a long time, the habit was to do 2 scans: a preview scan (low resolution) and then only do the real scan (high res). None of them were displayed in real time. Windows API (WIA) still tries to enforce that old way (which is *3%$@# annoying). And I guess some applications may still use this old way.

Basically, scanning is a mess, and cross-platform scanning even more. This is why I'm currently working on a [cross-platform scan library](https://doc.openpaper.work/libinsane/latest/doxygen/html/index.html).

Anyway, I would prefer an approach that is not dependent on any GUI at all. I could even add support for it to my library later on.
From what I see, portals require Dbus but there is no reason for them to require GTK or any other GUI framework, right ?

**Comment by hadess** [comment]
This bug isn't about having a single UI to import from scanners, and scanners have a lot more UI for advanced cases than you might think.

The current approach would be to allow applications (with filtering from a portal) within a Flatpak to access saned running on the host.

Whether somebody wants to then write a way to for apps to use scanners as easily as they would a file chooser is a problem for the toolkit to solve, not something we'd want at this point.

**Comment by jflesch** [comment]
> It might start with a Yes/No message dialog, of course :).

Yes, I think that for security reasons a dialog "allow X to access scanner Y ?" makes sense. But I don't think anything else deserve a specific GUI. Unlike the file system, when you have access only to a specific scanner, there isn't really a way to do much harm.
The only thing that can come to my mind is that some scanner can act also as file storage (you scan the image when your computer is off, and you get it from the scanner once it's one). But I'm not even sure whether Sane does support that feature.

Note that this reasoning may also apply to most other device types ;-)

**Comment by nedrichards** [comment]
For scanners, @hadess and @jflesch are right we don't need anything more than we need for pulseaudio. --socket=scanner or sane or whatever, but yes - this should either be a separate ticket or we should elbow @matthiasclasen out into his own ticket for the original point.

**Comment by matthiasclasen** [comment]
patches welcome

**Comment by leamas** [comment]
Perhaps a separate bug: serial ports. Used in applications like opencpn (serial GPS position data) and in any attempt to package the Arduino SDK.

Typically, serial ports are  660 and owned by root:dialout. With native packaging, it's reasonably simple to add the running user to e. g. the dialout group and thus have access to the device. Flatpak does not allow this, whatever permissions the native user has the sandbox permissions are unusable, even with full device access.

The only working solution I have at this point is to make the native device 666 using an udev rule. Needless to say, it's not satisfactory.  Seems like a "portal" is a possible better solution here, but to be honest I havn't really any idea what this actually means.

**Comment by leamas** [comment]
Perhaps a separate bug: serial ports. Used in applications like opencpn (serial GPS position data) and in any attempt to package the Arduino SDK.

Typically, serial ports are  660 and owned by root:dialout. With native packaging, it's reasonably simple to add the running user to e. g. the dialout group and thus have access to the device. Flatpak does not allow this, whatever permissions the native user has the sandbox permissions are unusable, even with full device access.

The only working solution I have at this point is to make the native device 666 using an udev rule. Needless to say, it's not satisfactory.  Seems like a "portal" is a possible better solution here, but to be honest I havn't really any idea what this actually means.

**Comment by hadess** [comment]
Scanner specific discussion moved to https://github.com/flatpak/xdg-desktop-portal/issues/218

**Comment by mwleeds** [comment]
Boxes [needs access to USB drives](https://gitlab.gnome.org/GNOME/gnome-boxes/issues/236), so I think that's also within the scope of this issue.

**Comment by hadess** [comment]
I'm going to close this bug. It's become a mish-mash of different use cases and device types being merged into one.

> Totem case: need full dvd access. Problem is that we can only set up device access at startup time, since we want to drop privileges. VLC also wants full optical device access, which is considered to be a showstopper for publishing VLC as a Flatpak.

This is block device level access. See #228 

> joystick: one idea would be to have the compositor forward events in a somewhat transparent way. This has the advantage that we can subject these events to focus. But there's an open question if we need any custom ioctls. Need to talk to domain experts: sdl and libinput maintainers.

We will likely implement this as a Wayland extension instead. Peter wrote a spec for it, Garnacho wrote a compositor-side implementation for mutter, and I volunteered to write the toolkit side of things.

This is the branch for mutter:
https://gitlab.gnome.org/GNOME/mutter/tree/wip/carlosg/inputfd

This is best discussed on the Wayland mailing-list, or on GNOME's gitlab.

> Richard wants to have usb access to update firmware. We tentatively agreed to just have a way to bind mount /dev in the metadata, for cases like this. Full device access is not quite as scary as it sounds since permissions still limit what is accessible.

See #227 

There's also #218 for scanners and #229 for serial ports.
